### ISSUE TITLE
Source Maps Should be Handled during the Build

### ISSUE THREAD

**Issue by narkisr** [ngbp/ngbp#180]
Using map files like:

``` javascript
 vendor_files: {
    js: [
      'vendor/angular/angular.js',
      'vendor/angular-animate/angular-animate.js',
      'vendor/underscore/underscore-min.js',
      'vendor/underscore/underscore-min.map',
      'vendor/underscore.string/dist/underscore.string.min.js',
      'vendor/angular-resource/angular-resource.min.js',
      'vendor/angular-resource/angular-resource.min.js.map',
      'vendor/angular-bootstrap/ui-bootstrap-tpls.min.js',
      'vendor/angular-ui-router/release/angular-ui-router.js',
      'vendor/angular-ui-utils/modules/route/route.js',
      'vendor/angular-growl/build/angular-growl.js',
      'vendor/angular-cookies/angular-cookies.js' ,
      'vendor/jquery/jquery.js'
    ],
```

Causes Uglify.js to fail when running grunt:

``` bash
Running "uglify:compile" (uglify) task
>> Uglifying source "bin/assets/ng-boilerplate-0.3.1.js" failed.
Warning: Uglification failed. Use --force to continue.

Aborted due to warnings.
```

**Comment by alex6o** [comment]
we had the same problem here with source maps; our solution is that we use only unminified versions of scripts and generate a source map for all compiled scripts.

**Comment by narkisr** [comment]
Yeah that is one solution I guess, still it would be nice to have some support in the current infra

**Comment by joshdmiller** [comment]
Does anyone have an idea of how to implement this? The vendor go through uglification, which you wouldn't want to do if you're going use the already-minified versions with their map files. There is probably a feature request or PR in here somewhere...

**Comment by rdthompson** [comment]
Josh,

After this AngularJS project I'm on which uses Parse.com, I may be able to provide some great contributions.

Keep it going man, it's a great boilerplate.  Have you heard of the new ng-hub?  Here,

http://www.angularjshub.com/examples/modules/

It would be great to get ng-boilderplate up there as a Premiere starting point for AngularJS integrating internationalization, real-time, cloud storage.

Can you please add me to the "Digest" for this project?

I want to stay on top of it as I love the project, but right now too many e-mails.

~ Robert

**Comment by ToucheSir** [comment]
I've currently been using a rather hackish patch in the gruntfile's <code>index</code> task so that files are referenced by their original path instead of their location after copying. Not sure if that completely breaks compiling, however....

**Comment by joshdmiller** [comment]
Just to clarify, the source maps shouldn't be in the `vendor_js` array at all as, unless I'm mistaken, the browser will load it from the specified path, if necessary. Therefore, source maps should exist _alongside_ the scripts but not concatenated _with_ scripts.

But anyway, I think my perspective here is that we need to distinguish between build/development, compile, and production environments. In build/development, not only should the map files not be included, but the minified files should be avoided too; even with the maps, it's a lot easier to find/track issues and bugs with the unminified sources. In production, we shouldn't include _any_ vendor files if we can avoid it because they should be properly served from CDN.

So the only place where map files should be an issue is in a pre-production compile, right? And how important is it to include map files at that point?

Also, FWIW, v0.4 of ngbp will have CDN functionality that will map source paths to target paths:

``` js
// Gruntfile.js
grunt.initConfig({
  cdn: {
    vendor: {
      "jquery/jquery.js": "//cdn.whatever.com/2.0.3/jquery.min.js",
      "angular/angular.js": "//cdn.whatever.com/1.2.0/angular.min.js"
    }
  }
});
```

The source map specified in the scripts will be appropriately included by the browser, if supported. The issue will still exist without the CDN, however, as the maps can exist neither inside the source nor alongside by default in ngbp. A new source maps task may be necessary for the new ngbp v0.4.

What are your thoughts, gents?

@rdthompson I am not familiar with ng-hub yet. Who runs it?

**Comment by ToucheSir** [comment]
@joshdmiller The sourcemaps must be in the vendor_files array so that they may be copied over with the scripts. This is necessary because the sourcemap urls in the script files use relative urls, so browsers freak out and are unable to determine the correct path. 

Also, in regards to sourcemaps, is there any plan on integrating LESS sourcemap support? I've had to switch out grunt-recess for grunt-contrib-less because recess doesn't have sourcemap functionality.

**Comment by joshdmiller** [comment]
@ToucheSir But the `vendor_js` array **isn't** to copy static vendor files; `vendor_assets` does that. `vendor_js` gets processed (linting, unit test inclusion, minifying, etc); `vendor_assets` does not. Your point about the relative URLs was exactly my point, but also my point was that we don't want source maps to be processed at all, which is why I said it will probably require a new task, but that it shouldn't be at all relevant for the _build_ - only the compile. Again, this is because minified files shouldn't be put through `vendor_js` either for the same reason - they shouldn't be reprocessed. Or am I missing something?

ngbp v0.4 switched to `grunt-contrib-less` for other reasons, so adding sourcemap support shouldn't be too hard, but it's not on the pipeline just yet. v0.4 is a pretty big shakeup and as you may have seen on other issues here I am planning a developer preview in about a week to test the waters; assuming it lands okay, the next step is to start thinking about things like this, which will be a little trickier I think - I encountered some difficulty already with special requirements for Jade.

If you're interested in helping out developing the remaining features of v0.4 (and if you have the time), shoot me an email - we could use your help.

**Comment by ToucheSir** [comment]
@joshdmiller Wow, I completely disregarded the fact that vendor_assets exists...you're absolutely right. Perhaps vendor assets could be separated into unminified and minified categories? Currently at a loss as to how that would work with concatenation and uglification. 

I'm afraid school will prevent me from finding enough time for helping much with 0.4, but looking forward to it nonetheless!

**Comment by joshdmiller** [comment]
@ToucheSir Yeah, it will definitely take more thought. But source maps are only getting bigger, so it's something we'll definitely need to solve. Once I get v0.4 out, I hope that between you, me, and everyone else that someone will be struck with a stroke of genius and the problem will be resolved. :-)

In the meantime, I'm going to change the subject of this issue to something more apropos and leave it open for the time being. Feel free to shoot your ideas, folks! (but honestly don't bother with a v0.3 PR for this as it will be unportable to v0.4. If anyone wants to solve this in v0.4, shoot me an email.)

**Comment by alex6o** [comment]
hi! with the 1.6.1 release of coffeescript the generation of sourcemaps is supported; maybe this could be more important for developing angular apps than the sm of vendor files? are there any plans to take these sm into account?

**Comment by chapati23** [comment]
just ran into this issue, too. i don't know if it's solvable with the current project structure. the problem is that the generated stylesheets end up in /build/assets (in my case) and the source (sass-)files live in /src/sass.

because the index.html lives in /build the browser has no possibility to find the correct source because he can't step out of /build and into /src (where the sass-files are the sourcemap is referring to)

my workaround at the moment is to copy all my sass-files into /build/src/sass but this is ugly. also, i can't use chrome's awesome workspaces feature for live-editing with this solution.

any ideas?

**Comment by chapati23** [comment]
just thought about it again and found a better workaround: i created a symlink vom /build/src/sass to /src/sass — that did the trick. sourcemaps and chrome workspaces working now :)
